File: app/src/main/java/com/yechaoa/materialdesign/activity/TextInputLayoutActivity.java
package com.yechaoa.materialdesign.activity;

import android.support.design.widget.TextInputEditText;
import android.support.design.widget.TextInputLayout;
import android.text.Editable;
import android.text.TextWatcher;

import com.yechaoa.materialdesign.R;

import butterknife.BindView;

public class TextInputLayoutActivity extends ToolbarActivity {

    @BindView(R.id.til_name)
    TextInputLayout mTilName;
    @BindView(R.id.til_password)
    TextInputLayout mTilPassword;
    @BindView(R.id.et_name)
    TextInputEditText mEtName;

    @Override
    protected int getLayoutId() {
        return R.layout.activity_text_input_layout;
    }

    @Override
    protected void setToolbar() {
        mToolbar.setTitle(R.string.text_input_layout);
    }

    @Override
    protected void initView() {
        mEtName.addTextChangedListener(mTextWatcher);
    }

    private TextWatcher mTextWatcher = new TextWatcher() {
        @Override
        public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

        }

        @Override
        public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {

        }

        @Override
        public void afterTextChanged(Editable editable) {
            if (mTilName.getEditText().getText().length() > mTilName.getCounterMaxLength())
                mTilName.setError("输入内容超过上限");
            else
                mTilName.setError(null);
        }
    };

}

File: app/src/main/java/com/yechaoa/materialdesign/activity/TextInputLayoutActivity.kt
package com.yechaoa.materialdesign.activity

import android.text.Editable
import android.text.TextWatcher
import com.yechaoa.materialdesign.R
import com.yechaoa.materialdesign.databinding.ActivityTextInputLayoutBinding

class TextInputLayoutActivity : ToolbarActivity<ActivityTextInputLayoutBinding>() {

    override fun getViewBinding(): ActivityTextInputLayoutBinding {
        return ActivityTextInputLayoutBinding.inflate(layoutInflater)
    }

    override fun setToolbar() {
        mToolbar.setTitle(R.string.text_input_layout)
    }

    override fun initView() {
        mBinding.etName.addTextChangedListener(mTextWatcher)
    }

    private val mTextWatcher: TextWatcher = object : TextWatcher {
        override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
        override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
        override fun afterTextChanged(editable: Editable) {
            if (mBinding.tilName.editText!!.text.length > mBinding.tilName.counterMaxLength)
                mBinding.tilName.error = "输入内容超过上限"
            else
                mBinding.tilName.error = null
        }
    }

}